File: 3488.closest-equal-element-queries.cpp
#
# @lc app=leetcode id=3488 lang=cpp
#
# [3488] Closest Equal Element Queries
#
# @lc code=start
class Solution {
public:
    vector<int> solveQueries(vector<int>& nums, vector<int>& queries) {
        unordered_map<int, vector<int>> index_map;
        int n = nums.size();
        // Fill hashmap with indices for each value in nums
        for (int i = 0; i < n; ++i) {
            index_map[nums[i]].push_back(i);
        }
        vector<int> result;
        // Process each query
        for (int query : queries) {
            int target = nums[query];
            if (index_map[target].size() == 1) { // Only one occurrence of target value
                result.push_back(-1);
                continue;
            }
            int min_distance = INT_MAX;
            // Find minimum distance in a circular manner
            for (int idx : index_map[target]) {
                if (idx != query) { // Don't compare with itself
                    int dist_cw = (idx - query + n) % n; // Clockwise distance
                    int dist_ccw = (query - idx + n) % n; // Counter-clockwise distance
                    min_distance = min(min_distance, min(dist_cw, dist_ccw));
                }
            }
            result.push_back(min_distance);
        } 
        return result; 
    } 
}; 
# @lc code=end

File: 3488.closest-equal-element-queries.java
#
# @lc app=leetcode id=3488 lang=java
#
# [3488] Closest Equal Element Queries
#

# @lc code=start
class Solution {
    public List<Integer> solveQueries(int[] nums, int[] queries) {
        Map<Integer, List<Integer>> indexMap = new HashMap<>();
        int n = nums.length;
        
        // Preprocess: map each number to its list of indices
        for (int i = 0; i < n; i++) {
            indexMap.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }
        
        List<Integer> result = new ArrayList<>();
        // Process each query
        for (int query : queries) {
            int target = nums[query];
            List<Integer> indices = indexMap.get(target);
            if (indices == null || indices.size() == 1) { // No other occurrence or only one occurrence at query itself
                result.add(-1);
            } else {
                int minDist = Integer.MAX_VALUE;
                // Find minimum distance considering circularity
                for (int idx : indices) {
                    if (idx != query) {
                        int dist = Math.min(Math.abs(idx - query), n - Math.abs(idx - query)); // Circular distance calculation 
                        minDist = Math.min(minDist, dist); 
                    }
                }   
                result.add(minDist); 
            }    
        }   
        return result;  
    }   
}
# @lc code=end